Merging `maybeOverlappingIndices` and `has_internal_overlap`

jjsjann123 opened the pull request:
These two functions basically serves the same purpose.
`has_internal_overlap` provides a more flexible API (where
`MemOverlap::TOO_HARD` is provided), while `maybeOverlappingIndices` provides
meaningful flags on more general cases.

We get the best out of each and unifies the API.

Test cases are also provided. One thing to note here is, we should NEVER assert
on return being `MemOverlap::TOO_HARD`, which would block future improvement to
the `has_internal_overlap` function to cover more grounds.
We should instead assert on not returning the incorrect flags for cases not yet
supported by current implementation.

jjsjann123 committed:
[Merging `maybeOverlappingIndices` and `has_internal_overlap`]

These two functions basically serves the same purpose.
`has_internal_overlap` provides a more flexible API (where
`MemOverlap::TOO_HARD` is provided), while `maybeOverlappingIndices` provides
meaningful flags on more general cases.

We get the best out of each and unifies the API.

Test cases are also provided. One thing to note here is, we should NEVER assert
on return being `MemOverlap::TOO_HARD`, which would block future improvement to
the `has_internal_overlap` function to cover more grounds.
We should instead assert on not returning the incorrect flags for cases not yet
supported by current implementation.

SsnL commented:
could https://github.com/pytorch/pytorch/blob/3fe00f0c906ab4166058a8619b3c2edf121b2537/tools/autograd/templates/Functions.cpp#L1387
be also merged by any chance? I remember @zou3519 telling me that it does a different thing though...

jjsjann123 commented:
The implementation is slightly different, but I think it's doing the same thing. 
The implementation here didn't special case tensor with a dimension of size 0 here.

Just to be on the safe side, let me take a careful look at that `NOTE [ Detecting Memory Overlap Within A Strided Tensor ]` before merging it :)

VitalyFedyunin commented on aten/src/ATen/MemoryOverlap.cpp:
```diff
@@ -3,24 +3,83 @@
 }
 
+/*
+Returns false if there is no possibility that the tensor
```
Returns `MemOverlap::NO`

VitalyFedyunin commented on aten/src/ATen/MemoryOverlap.cpp:
```diff
@@ -3,24 +3,83 @@
-    return MemOverlap::YES;
+  std::vector<SizeAndStride> info(t->dim());
+
+  int dims = t->dim();
```
dims <= 1 tensor can't have overlaps, right?

zou3519 replied:
`x = torch.tensor(3.).expand(100)` has dim 1 and internal memory overlap

VitalyFedyunin replied:
(facepalm)

VitalyFedyunin commented on aten/src/ATen/cuda/CUDAApplyUtils.cuh:
```diff
@@ -645,7 +646,7 @@ inline bool CUDA_tensor_apply1(at::Tensor a,
   Since this is a unary op, we can easily first check for expanded dimensions
   (with stride 0), and remove them, to avoid calling .contiguous() in such
-  case when detail::maybeOverlappingIndices(a) returns true.
+  case when has_internal_overlap(a) returns true.
```
returns `MemOverlap::NO`

apaszke commented on aten/src/ATen/MemoryOverlap.cpp:
```diff
@@ -3,24 +3,83 @@
+  }
+
+  /* Ascending order (innermost dimension in sorted view is at [0]) */
+  qsort(info.data(), nonSize1Dims, sizeof(SizeAndStride), compareSizeAndStride);
```
Ugh, can we use `std::sort` instead of `qsort`? That has a better chance of e.g. inlining the comparator.

jjsjann123 commented:
Realized the logic in the function i merged earlier was wrong. It should have be doing the same thing that @SsnL pointed out.

Let me merge all of them together.

SsnL commented:
For the function I referenced earlier, it is kind of tricky because part of the algorithm is implemented in another function at 
https://github.com/pytorch/pytorch/blob/3fe00f0c906ab4166058a8619b3c2edf121b2537/tools/autograd/templates/Functions.cpp#L1428

The full algorithm is described at
https://github.com/pytorch/pytorch/blob/3fe00f0c906ab4166058a8619b3c2edf121b2537/tools/autograd/templates/Functions.cpp#L1190-L1201

The rest of that long NOTE is just proofs on how the property holds for all **current** strided tensor view ops, except for `as_strided` and `unfold`.

Since the implemented is split into two functions (for perf reasons), feel free to skip merging that if it becomes too hard. I can give it a try later. But if you are going to follow the same (or a similar algorithm), I hope you can move the note to the new place. Thank you. :)

ezyang commented:
@VitalyFedyunin @SsnL  are there any further blockers on this?

VitalyFedyunin commented:
@ezyang my comments not addressed

ezyang reviewed (changes requested):
comments are not addressed

jjsjann123 commented:
Simon's request is not address neither. I'm currently occupied on other stuff right now, I'll get back to this at some point and put an update when it's ready for another review. Apology for the delay 😢

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into assert_PR

jjsjann123 committed:
[addressing review comments]

further merge /pytorch/tools/autograd/templates/Functions.cpp so we have only
one implementation of memory overlapping check.

jjsjann123 commented:
Sorry for the delayed response. But address the request from @SsnL and typos that @VitalyFedyunin got.

jjsjann123 committed:
fixing a typo which results in test failure

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into assert_PR

VitalyFedyunin commented:
Hi! Can you please rebase and attach Before/After benchmarks, thanks.

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into assert_PR

jjsjann123 committed:
merging overlap tests to memory_overlapping_test.
fixing overlapping check:
1. return MemOverlap::YES on case where a dimension has
   [size > 1 && stride <= 0]
2. return MemOverlap::NO on case where a dimension has [size == 0]

jjsjann123 commented:
Two things this PR does:

1. The main idea is to fix: https://github.com/pytorch/pytorch/blob/949678bd9e82c12efe77fb0fb5471a17ded128f4/aten/src/ATen/cuda/detail/IndexUtils.cu#L64-L68
`maybeOverlappingIndices` here is basically wrong. It would return `false` with tensor (size: [2, 2, 10], stride: [11, 10, 1]), while the overlapping could be detected using `has_internal_overlap`;

2. Improve the check for `has_internal_overlap`: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/MemoryOverlap.cpp#L10-L24
Pushing more cases from `TOO_HARD` to `NO` and `YES` repectively.
e.g.
```
x = torch.randn(4)

# output has been updated from `TOO_HARD` to `NO`
print(torch._debug_has_internal_overlap(x[::2]))

# output has been updated from `TOO_HARD` to `YES`
z = x.as_strided((2, 2), (1, 1))
```

jjsjann123 committed:
fixing mistakes in code merging

jjsjann123 commented:
So something naive as:
```
  1 import torch
  2 import timeit
  3
  4 x = torch.randn(128, 64, 128, 128)
  5
  6 setup = """
  7 import torch
  8 x = torch.randn(128, 64, 128, 128)
  9 """
 10 stmt = """
 11 torch._debug_has_internal_overlap(x[::2, ::2, ::2, ::2])
 12 torch._debug_has_internal_overlap(x[:, ::2, ::2, ::2])
 13 torch._debug_has_internal_overlap(x.as_strided((2,2,2),(1,1,1)))
 14 """
 15
 16 print(timeit.timeit(stmt=stmt, setup=setup, number=10000))
```
gives ~0.140 on my current branch.
Same code gives ~0.122 on docker hub 1.3-runtime (14 days old `c3f6590d7263`) & ~0.142 on docker hub 1.3-cuda10.1-cudnn7-devel (14 days old `d22bdcc867e2`)

I'll be surprised that simple host code like this would have impact on performance.

Meanwhile, it did improve our error checking on memory overlap.
```
import torch
x = torch.zeros(3)
y = x.as_strided((2, 2), (1, 1))

# It now throws an error for multiple element being written-to.
# which is desired behavior for tensor iterator
y += 1
```

jjsjann123 commented:
Failure from compatibility test:

```Oct 29 03:17:23 processing existing schema:  aten::full_like(Tensor self, Scalar fill_value, *, int? memory_format=None) -> (Tensor)
Oct 29 03:17:23 Can NOT find backward compatible schemas after changes for schema aten::full_like(Tensor self, Scalar fill_value, *, int? memory_format=None) -> (Tensor) from the following candidates:
Oct 29 03:17:23 [
Oct 29 03:17:23 aten::full_like(Tensor self, Scalar fill_value) -> (Tensor)
Oct 29 03:17:23 	aten::full_like.dtype(Tensor self, Scalar fill_value, *, int dtype, int layout, Device device, bool pin_memory=False) -> (Tensor)
Oct 29 03:17:23 ]
Oct 29 03:17:23 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.```

Looks like something related to previous NHWC changes and not related to my change.

jjsjann123 commented:
@VitalyFedyunin Requests have been addressed. Please take another look :bowing_man:

jjsjann123 commented:
I can try to resolve the conflicts again. Is there any remaining issue that is blocking the merge? I'd like to resolve that when merging as well.

VitalyFedyunin commented:
Please hold on merge, I will do my best to provide you with detailed review by Monday (Nov 18).
